docs/preview-builds.md

```diff
@@ -0,0 +1,33 @@
+# Testing monorepo changes against other projects
+
+Let's say that you're working on a feature in one of our products and this feature relies on changes that you need to make to one or more packages within this repo. You make those changes here, but now you want to make sure that they fulfill the need you have in your product and don't break anything. How do you integrate these changes into your product?
+
+Previously you might have used the "commit id" method to solve this problem, where you:
+
+- push up a branch in this repo
+- copy the commit id of the latest commit on that branch
+- open `package.json` in your project
+- replace the version of some package with `MetaMask/controllers#<commit id>`
+
+That method won't work anymore (at least until we have all repos using Yarn v3). So now what do you do?
+
+Enter **preview builds**. These are ephemeral versions of packages which are built on demand and published to GitHub Package Registry.
+
+To use them, you'll need to do a few of things:
+
+1. You'll need to create a personal access token within your GitHub account. If you have not already done so, go into the [token settings for your account](https://github.com/settings/tokens) and [create a token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic). (Note: there are two types of tokens; you'll want a **classic** token.)
+2. In your product, you'll need to add an `.npmrc` file with the following content:
+
+   ```
+   @metamask:registry=https://npm.pkg.github.com
+
+   //npm.pkg.github.com/:_authToken=<insert your personal access token here>
+   ```
+
+3. Over here in `controllers`, push up a draft PR for this repo (or place an existing PR into draft mode), then make a comment on the PR with the text:
```

Comment:
We should probably make it normative that people run CI on their own branches? (As-is this not explicit about it, considering "this repo" technically depends on if you read it before or after forking). Consider that this will also be perused by people unaffiliated with the MetaMask team.
Would there be additional setup instructions (like [here](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#managing-github-actions-permissions-for-your-repository) we should mention or link to to capture that?

```suggestion
3. In you fork of `MetaMask/controllers`, create a draft PR for this repo (or place an existing PR into draft mode), then make a comment on the PR with the text:
```

Reply:
This trigger only works for branches on this repo. Forks aren't supported by preview builds. We have been encouraging people on the team _not_ to use forks, because compatibility with certain CI steps is worse.

It's difficult to safely build any integrations that work with forks _and_ allow interacting with the upstream repo. We don't want to give untrusted third parties access to a write token for this repository for example, even one scoped just to comments, because that would let them create fake `metamaskbot` comments.

Reply:
I guess I have to look into preview builds and try it out to see how the limitations arise.

Some of the below is just opinion, even if I write normatively.

Without having seen more than 1 case of things behaving differently, it seems like when it comes to producing builds and uploading them, there is fundamentally nothing apart from setting different registry and access token etc that should differ...? Ie everything done in this central repo should be doable transparently on forks as long as the expected env vars / secrets are there.

This is also with the assumption that we want to allow contributors on equal terms even if they're not part of the MM org. Even if we'd say and think we do, failing CI steps, as well as documentation presented as general but effectively only relevant for MM org internal use (I believe that first-time contributors need manual approval to have their draft PRs builds, and that having a draft PR pop up here for everyone who want to make a build is going to become untenable sooner than later) make that less likely and affect expectations, reducing probability of potential contributors actually engaging constructively.

Consider also overage costs and throughput for CI - things add up.

I completely understand that documentation should reflect reality. But I also would like to accommodate and normalize a more distributed flow. Perhaps this file/section can stay on the cooler (that is, pull it out of this PR for now to not block) until the outstanding CI issues are resolved and we can present something that can also set best practices and be more generally applicable?

Until then (given some of the complications above), and without clearer enumeration of alternatives (talked about below), I'd consider this internal docs rather than something that's useful to make public.

Reply:
One question I have here as I rewrite this is, why are we requiring the `@metamask-bot` comment to only work on draft PRs? I feel like it complicates things a bit because I have to include this bit:

> Continue to keep your `controllers` PR in draft mode until you're sure that you don't need to make any new changes. If you do push up a new commit to your PR, however, simply post another `@metamask-bot preview-build` comment and this will trigger another build.

Reply:
I don't think we are doing that. We used the preview-builds command for the v34 release PR well after bringing it out of draft.

We're requiring extension and mobile PRs that utilize these builds to be draft PRs, so that we don't accidentally merge preview builds into the main branch.